nixos/actual: add environmentFile option

gepbird opened the pull request:
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform:
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [x] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [ ] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [ ] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [x] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

gepbird committed:
nixos/actual: add environmentFile option

pyrox0 commented:
LGTM, though not sure why I was requested as a reviewer, since I'm not a maintainer of the package or module.

oddlama commented:
> Thanks, the reason I requested a review from you is because you merged a PR that had a similar topic in #440038, and the maintainers didn't show up for two weeks, and it's a quick to review change.

I'm sorry, but neither @PatrickDaG nor I were requested for review on this PR until just **3 hours** before this was merged. I think I speak for both of us when I say that we do our best in maintaining our packages and modules. Unfortunately, in this short time-span neither of us had a chance to even notice it, since we both have to catch up on all of this stuff after work hours. (Unfortunately we cannot monitor all incoming PRs)

Now unfortunately, I would actually have had some thoughts about this. I think this could have been solved better with `utils.genJqSecretsReplacementSnippet` which is a lot more flexible, instead of introducing the harder to use `environmentFile` option. So I would actually be in favor of reverting this but due to the upcoming 25.11 release we are now currently in a phase that restricts breaking changes - and technically reverting would now constitute a breaking change.

Now we could still switch to the aforementioned utility while keeping the introduced environmentFile option, but IMO this now forces us into a potentially suboptimal solution.

What are your thoughts on this @PatrickDaG @gepbird @pyrox0  ?

pyrox0 commented:
It was confirmed by release managers that the option would be fine to remove in favor of an improved interface. Please request me for review on the revert PR.

gepbird commented:
> I'm sorry, but neither @PatrickDaG nor I were requested for review on this PR until just 3 hours before this was merged. I think I speak for both of us when I say that we do our best in maintaining our packages and modules. Unfortunately, in this short time-span neither of us had a chance to even notice it, since we both have to catch up on all of this stuff after work hours. (Unfortunately we cannot monitor all incoming PRs)

You're absolutely right, my apologies. I just looked at the PR being open for 2 weeks and the awaiting review next to your names, I forgot that the automatic reviewer request doesn't work for NixOS module and I manually requested your review.

> Now unfortunately, I would actually have had some thoughts about this. I think this could have been solved better with utils.genJqSecretsReplacementSnippet which is a lot more flexible, instead of introducing the harder to use environmentFile option. So I would actually be in favor of reverting this but due to the upcoming 25.11 release we are now currently in a phase that restricts breaking changes - and technically reverting would now constitute a breaking change.

I'm happy to hear there's a standard and more easy to use `genJqSecretsReplacementSnippet` utility and I'm surprised there are not many uses of it considering it's 6 years old. In https://github.com/NixOS/nixpkgs/pull/395551 I implemented a very similar secret management solution, but the downside was that we ended up with many secret files (at least with `agenix`, I'm sure it's better with `sops-nix`), and when the upstream added environment variable support, I added an option for it which we are using. Currently both options exist for giving secrets to the module.

I think for `actual` we can definitely use `utils.genJqSecretsReplacementSnippet`, but some people may prefer using an environment file instead and I don't see a big problem or maintainance burden of keeping both.

I don't have that strong of an opinion on it, if you think only using `utils.genJqSecretsReplacementSnippet` is the best, it's fine by me and feel free to revert this.

oddlama commented:
> You're absolutely right, my apologies. I just looked at the PR being open for 2 weeks and the awaiting review next to your names, I forgot that the automatic reviewer request doesn't work for NixOS module and I manually requested your review.

All good, stuff happens.

> I'm happy to hear there's a standard and more easy to use `genJqSecretsReplacementSnippet` utility and I'm surprised there are not many uses of it considering it's 6 years old.

Definitely needs to become more discoverabe!

> I think for `actual` we can definitely use `utils.genJqSecretsReplacementSnippet`, but some people may prefer using an environment file instead and I don't see a big problem or maintainance burden of keeping both.

Definitely, but I also think we don't need it for actual. Btw `utils.genJqSecretsReplacementSnippet` can be used to hide full JSON subtrees, but that's of course not very practicable if there are secrets in many different places in the json.

> I don't have that strong of an opinion on it, if you think only using `utils.genJqSecretsReplacementSnippet` is the best, it's fine by me and feel free to revert this.

If it doesn't cause to much inconvenience I'd love to go the genJqSecretsReplacementSnippet route.
